core/tx_pool.go

```diff
@@ -680,6 +685,29 @@ func (pool *TxPool) add(tx *types.Transaction, local bool) (bool, error) {
 			pool.removeTx(tx.Hash(), false)
 		}
 	}
+	// If there isn't enough free size in pool, discard underpriced transactions
+	if pool.all.TotalSize() + tx.Size() > pool.config.TotalPoolSizeLimit {
+		sizeToFree := pool.all.TotalSize() + tx.Size() - pool.config.TotalPoolSizeLimit
+		drop, freedSize := pool.priced.FreeSize(sizeToFree, tx.GasPrice(), local, pool.locals)
+		totalFreeSize := pool.config.TotalPoolSizeLimit - pool.all.TotalSize() + freedSize
+		if(totalFreeSize < tx.Size()){
```

Comment:
A curious corner case not handled is if all the transactions are local and the memory limit is exceeded. The invariant currently is that no local transaction is ever dropped. We need to retain that. Even if there's no space for a tx, as long as it's local, we need to store it. If we go OOM, so be it.